### DISCUSSION TITLE
Question: Authorization token

### DISCUSSION THREAD

**Discussion by Mister-Ritom** [pocketbase/pocketbase#3222] [Q&A] [answered]
How does the authoriztion token work before the user signs in. I am trying to create a pocketbase sdk for android and having some problems with the authorization

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand what do you mean. There is no token if the user isn't authenticated.

The auth token can be generated by using the following APIs:

- Auth record token via:
  - password: https://pocketbase.io/docs/api-records/#auth-with-password
  - OAuth2: https://pocketbase.io/docs/api-records/#auth-with-oauth2
  - refresh: https://pocketbase.io/docs/api-records/#auth-refresh

- Admin token via:
  - password: https://pocketbase.io/docs/api-admins/#auth-with-password
  - refresh: https://pocketbase.io/docs/api-admins/#auth-refresh
  
And then you need to add the generated token from above as `Authorization: YOUR_TOKEN` header to each of your requests.

Keep in mind that the PocketBase APIs are stateless and there no cookies/sessions. The only requirement to be considered authenticated is to have a valid `Authorization` header with the request.

**Reply by ganigeorgiev** [reply]
_You can use the Dart SDK as a reference implementation as it should be pretty close to Kotlin/Java - https://github.com/pocketbase/dart-sdk (the only slightly more difficult thing I guess you can stumble on would be the realtime subscriptions but you can always fallback to manual SSE message parsing [as we do in the Dart SDK](https://github.com/pocketbase/dart-sdk/blob/master/lib/src/sse/sse_client.dart) if a suitable SSE client library is not found)_

**Comment by Mister-Ritom** [comment]
This is my code in kotlin with OkHttp
```
    fun PostRequest() {

        val body:HashMap<String, Any> = HashMap()
        body["name"] = "ritom"
        body["username"] = "ritom"
        body["password"] = "ritom12356"
        body["passwordConfirm"] = "ritom12356"
        body["email"] = "mail@email.com"
        val gson = Gson()
        val json = gson.toJson(body)
        val requestBody = json.toRequestBody("application/json".toMediaTypeOrNull())

        val request = Request.Builder()
            .url("http://127.0.0.1:8090/api/collections?collectionId=_pb_users_auth_")
            .post(requestBody)
            .addHeader("Content-Type", "application/json")
            .addHeader("AcceptLanguage", "`en-US")
            .addHeader("Authorization","")
            .build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw Exception("Unexpected code $response")
            println(response.body!!.string())
        }
    }
```

I get this exception
`Exception in thread "main" java.lang.Exception: Unexpected code Response{protocol=http/1.1, code=401, message=Unauthorized, url=http://127.0.0.1:8090/api/collections?collectionId=_pb_users_auth_}`

What am I doing wrong?

**Reply by ganigeorgiev** [reply]
The url you  are using is incorrect.
For authentication, the url should be `/api/collections/_pb_users_auth_/auth-with-password`.
For "user create" the url should be `/api/collections/_pb_users_auth_/records`.

The body fields also doesn't seem correct, if it is for authentication.

For more details about the API please refer to the above doc links.

**Reply by ganigeorgiev** [reply, marked as answer]
If you are trying to send a "user create" request, please check https://pocketbase.io/docs/api-records/#create-record (you can also find individual collection API info from the _Admin UI > {collection} > API Preview panel_).

**Comment by Mister-Ritom** [comment]
It worked thanks
